File: api/src/main/java/com/nfthub/api/controller/TagController.java
package com.nfthub.api.controller;

import com.nfthub.core.response.TagResponse;
import com.nfthub.core.service.TagService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.Parameters;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.tags.Tag;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@Tag(name = "Tag api")
@RestController
@RequestMapping("/api/v1/tags")
@RequiredArgsConstructor
public class TagController {
    private final TagService tagService;

    @Operation(summary = "태그 리스트 조회")
    @GetMapping
    @Parameters(
            value = {
                    @Parameter(
                            name = "tag",
                            description = "검색어를 통해 태그 조회",
                            schema = @Schema(type = "string")
                    )
            }
    )
    public List<TagResponse> getTagsByKeyword(
            @RequestParam(required = false) String keyword
    ) {
        return tagService.getTagsResByLikeKeyword(keyword);
    }

    @Operation(summary = "태그 단일 조회")
    @GetMapping("/{id}")
    public TagResponse getTagById(
            @PathVariable Long id
    ) {
        return tagService.getTagResByIdOrThrow(id);
    }

    @Operation(summary = "태그 이름으로 조회, 정확히 일치하는 태그 조회")
    @GetMapping("/name")
    public TagResponse getTagByName(
            @RequestParam String name
    ) {
        return tagService.getTagResByNameOrThrow(name);
    }

    @Operation(summary = "키워드로 태그 검색, like 검색 지원")
    @GetMapping("/search")
    public List<TagResponse> getTagsLikeKeyword(
            @RequestParam String keyword
    ) {
        return tagService.getTagsResByLikeKeyword(keyword);
    }
}

File: src/main/kotlin/com/nfthub/api/controller/TagController.kt
package com.nfthub.api.controller

import com.nfthub.api.service.TagService
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.tags.Tag
import org.springframework.web.bind.annotation.*

@Tag(name = "tag api")
@RestController
@RequestMapping("/api/v1/tags")
class TagController(
    private val tagService: TagService
) {
    @Operation(summary = "키워드 리스트 조회")
    @GetMapping
    fun getTags() = tagService.getTagResponses()

    @Operation(summary = "키워드 단일 조회")
    @GetMapping("/{tagId}")
    fun getTag(
        @PathVariable tagId: Long
    ) = tagService.getTagOrThrow(tagId)
}

@Tag(name = "tag manager")
@RestController
@RequestMapping("/manage/v1/tags")
class ManageTagController(
    private val tagService: TagService
) {
    @Operation(summary = "키워드 생성", description = "어드민 전용")
    @PostMapping
    fun createTag(
        @RequestBody name: String
    ) = tagService.createTag(name)

    @Operation(summary = "키워드 수정", description = "어드민 전용")
    @PatchMapping("/{tagId}")
    fun updateTag(
        @PathVariable tagId: Long,
        @RequestBody name: String
    ) = tagService.updateTagName(tagId, name)

    @Operation(summary = "키워드 삭제", description = "어드민 전용")
    @DeleteMapping("/{tagId}")
    fun deleteTag(@PathVariable tagId: Long) =
        tagService.deleteTag(tagId)
}